File: python-reference/src/ft_cli/commands/active.py
from __future__ import annotations

import typer

from ft_cli import auth, config
from ft_cli.client import ApiError, Client
from ft_cli.commands.campus import resolve_campus_id
from ft_cli.render import active_table, err_console


def fetch_active_locations(client: Client, campus_id: int, limit: int) -> list[dict]:
    locations: list[dict] = []
    for loc in client.paginate(
        f"/v2/campus/{campus_id}/locations",
        page_size=100,
        **{"filter[active]": "true", "sort": "-begin_at"},
    ):
        locations.append(loc)
        if len(locations) >= limit:
            break
    return locations


def resolve_campus(client: Client, slug: str | None) -> tuple[str, int]:
    cfg = config.load_config()
    chosen = (slug or cfg.default_campus_slug or "").strip().lower()
    if not chosen:
        raise typer.BadParameter(
            "no campus provided and no default set. Try `42 campus set <slug>`."
        )
    campus_id = resolve_campus_id(client, chosen)
    if campus_id is None:
        raise typer.BadParameter(
            f"unknown campus slug: {chosen}. Try `42 campus list`."
        )
    return chosen, campus_id


def active(
    campus: str | None = typer.Option(
        None, "--campus", "-c", help="campus slug (defaults to your saved campus)"
    ),
    limit: int = typer.Option(50, "-n", "--limit", help="max rows to show"),
) -> None:
    """Show users currently logged in at the campus."""
    try:
        with Client() as c:
            slug, campus_id = resolve_campus(c, campus)
            locations = fetch_active_locations(c, campus_id, limit)
    except (auth.AuthError, ApiError) as e:
        err_console.print(str(e))
        raise typer.Exit(code=1)

    active_table(locations, title=f"Active now @ {slug}")

File: src/commands/active.ts
import { AuthError } from "../auth.js";
import { ApiError, paginate } from "../client.js";
import { activeTable, err, withSpinner } from "../render.js";
import { resolveCampus } from "./campus.js";

export async function fetchActiveLocations(
  campusId: number,
  limit: number,
): Promise<any[]> {
  const out: any[] = [];
  for await (const loc of paginate<any>(
    `/v2/campus/${campusId}/locations`,
    { "filter[active]": "true", sort: "-begin_at" },
    100,
  )) {
    out.push(loc);
    if (out.length >= limit) break;
  }
  return out;
}

export async function activeCmd(opts: {
  campus?: string;
  limit?: string;
}): Promise<void> {
  const limit = Math.max(1, Number(opts.limit ?? 50));
  try {
    const { slug, id } = await resolveCampus(opts.campus);
    const locs = await withSpinner(
      `fetching active users at ${slug}...`,
      () => fetchActiveLocations(id, limit),
    );
    activeTable(locs, `Active now @ ${slug}`);
  } catch (e) {
    if (e instanceof ApiError || e instanceof AuthError) {
      err(e.message);
      process.exit(1);
    }
    if (e instanceof Error) {
      err(e.message);
      process.exit(1);
    }
    throw e;
  }
}
